Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift

```diff
@@ -50,6 +50,18 @@ fileprivate func getTokens(between first: TokenSyntax, and second: TokenSyntax)
   return tokens
 }
 
+fileprivate func getIndentation(for token: TokenSyntax) -> Trivia {
+  var ancestor: Syntax = Syntax(token)
+  while let parent = ancestor.parent {
+    ancestor = parent
+    if let indentation = ancestor.leadingTrivia.indentation(isOnNewline: true), !indentation.isEmpty {
+      return indentation
+    }
+  }
```

Comment:
I think you want to walk `token.previousToken(viewMode: .sourceAccurate)` instead of walking the ancestors. 

For example consider

```swift
func
  test() {}
```

If you want to get the indentation of `(`, you wouldn’t get `  ` because `test` is not the first token in any of its ancestors.

Reply:
Here’s the example I was thinking of

```swift
func 
  test() {} var other: Int
```

Based on how your code reads, I was expecting that we get the indentation of the missing `;` between `}` and `var`, which I would expect to be `  ` because that’s the indentation of the line that the `;` appears on. But I think your implementation would just go one level up and return empty indentation because `func` isn’t indented.
That’s why I was expecting `getIndentation` to walk tokens backwards using `previousToken` instead of walking up the tree hierarchy. 

Behavior-wise, I think your implementation is good. Not indenting `var` is definitely the superior choice to adding two spaces in front of it. To marry the two ideas, I think `getIndentation` should walk tokens backwards but base the indentation that we want on the first token in the `CodeBlockItem` that has the missing `;`. Does this make sense?

---

Unrelated, I would rename `getIndentation` -> `indentation`. Sounds more swifty.

Reply:
Okay, I think I got it. 
But we have a thing, that we can't find the right indentation for `var x = i3️⃣ x = i + x4️⃣ return Float(x)`

Because the return will get the `indentation` from `x = i + x`

We could of course check if the leading trivia is empty.

Reply:
I was thinking more of something like the following. And then we could move the extension to SyntaxExtensions.swift.

```diff
diff --git a/Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift b/Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift
index a21e346f..674c00b1 100644
--- a/Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift
+++ b/Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift
@@ -50,21 +50,31 @@ fileprivate func getTokens(between first: TokenSyntax, and second: TokenSyntax)
   return tokens
 }
 
-fileprivate func indentation(for token: TokenSyntax) -> Trivia {
-  var ancestor: Syntax = Syntax(token)
-  while let parent = ancestor.parent {
-    ancestor = parent
-
-    if let codeBlockItemSyntax = ancestor.as(CodeBlockItemSyntax.self) {
-      return codeBlockItemSyntax.leadingTrivia.indentation(isOnNewline: true) ?? []
+extension TokenSyntax {
+  /// The indentation of this token
+  ///
+  /// In contrast to `indentation`, this does not walk to previous tokens to
+  /// find the indentation of the line this token occurs on.
+  private var indentation: Trivia? {
+    let previous = self.previousToken(viewMode: .sourceAccurate)
+    return ((previous?.trailingTrivia ?? []) + leadingTrivia).indentation(isOnNewline: false)
+  }
+
+  /// Returns the indentation of the line this token occurs on
+  var indentationOfLine: Trivia {
+    var token: TokenSyntax = self
+    if let indentation = token.indentation {
+      return indentation
+    }
+    while let previous = token.previousToken(viewMode: .sourceAccurate) {
+      token = previous
+      if let indentation = token.indentation {
+        return indentation
+      }
     }
 
-    if let memberDeclListItemSyntax = ancestor.as(MemberDeclListItemSyntax.self) {
-      return memberDeclListItemSyntax.leadingTrivia.indentation(isOnNewline: true) ?? []
-    }
+    return []
   }
-
-  return []
 }
 
 fileprivate extension TokenSyntax {
@@ -719,21 +729,24 @@ public class ParseDiagnosticsGenerator: SyntaxAnyVisitor {
         // Only diagnose the missing semicolon if the item doesn't contain any errors.
         // If the item contains errors, the root cause is most likely something different and not the missing semicolon.
         let position = semicolon.previousToken(viewMode: .sourceAccurate)?.endPositionBeforeTrailingTrivia
+        var fixIts: [FixIt] = [
+          FixIt(message: .insertSemicolon, changes: .makePresent(semicolon)),
+        ]
+        if let firstToken = node.firstToken(viewMode: .sourceAccurate),
+          let lastToken = node.lastToken(viewMode: .sourceAccurate) {
+          fixIts.insert(FixIt(
+            message: .insertNewline,
+            changes: [
+              .replaceTrailingTrivia(token: lastToken, newTrivia: lastToken.trailingTrivia + .newlines(1) + firstToken.indentationOfLine)
+            ]
+          ), at: 0)
+        }
+
         addDiagnostic(
           semicolon,
           position: position,
           .consecutiveStatementsOnSameLine,
-          fixIts: [
-            node.lastToken(viewMode: .sourceAccurate).map {
-              FixIt(
-                message: .insertNewline,
-                changes: [
-                  .replaceTrailingTrivia(token: $0, newTrivia: $0.trailingTrivia + .newlines(1) + indentation(for: $0))
-                ]
-              )
-            },
-            FixIt(message: .insertSemicolon, changes: .makePresent(semicolon)),
-          ].compactMap { $0 },
+          fixIts: fixIts,
           handledNodes: [semicolon.id]
         )
       } else {
@@ -1180,21 +1193,23 @@ public class ParseDiagnosticsGenerator: SyntaxAnyVisitor {
         // Only diagnose the missing semicolon if the decl doesn't contain any errors.
         // If the decl contains errors, the root cause is most likely something different and not the missing semicolon.
         let position = semicolon.previousToken(viewMode: .sourceAccurate)?.endPositionBeforeTrailingTrivia
+        var fixIts: [FixIt] = [
+          FixIt(message: .insertSemicolon, changes: .makePresent(semicolon)),
+        ]
+        if let firstToken = node.firstToken(viewMode: .sourceAccurate),
+          let lastToken = node.lastToken(viewMode: .sourceAccurate) {
+          fixIts.insert(FixIt(
+            message: .insertNewline,
+            changes: [
+              .replaceTrailingTrivia(token: lastToken, newTrivia: lastToken.trailingTrivia + .newlines(1) + firstToken.indentationOfLine)
+            ]
+          ), at: 0)
+        }
         addDiagnostic(
           semicolon,
           position: position,
           .consecutiveDeclarationsOnSameLine,
-          fixIts: [
-            node.lastToken(viewMode: .sourceAccurate).map {
-              FixIt(
-                message: .insertNewline,
-                changes: [
-                  .replaceTrailingTrivia(token: $0, newTrivia: $0.trailingTrivia + .newlines(1) + indentation(for: $0))
-                ]
-              )
-            },
-            FixIt(message: .insertSemicolon, changes: .makePresent(semicolon)),
-          ].compactMap { $0 },
+          fixIts: fixIts,
           handledNodes: [semicolon.id]
         )
       } else {

```